### ISSUE TITLE
Jekyll 4.0 Ideas

### ISSUE THREAD

**Issue by oe** [jekyll/jekyll#6948]
:wave: Hello everyone! Summer is coming, and so is the implementation period for Jekyll 4.0. That's right, it's time to get some breaking changes in. To accommodate for this, we're opening this issue to collect interesting ideas that people would like to see implemented in 4.0. __Keep in mind that even if an idea receives a lot of support, there's no guarantee that it'll get implemented__ — that depends on if someone is free to actually implement it. We're all volunteers here, keep that in mind.

Feel free to revive old feature requests, too, just not something that we've explicitly rejected.

For an organized view of how we're consolidating ideas and features, check out our Project board: https://github.com/jekyll/jekyll/projects/2

**Comment by laukstein** [comment]
I vote for #6293 Markdown links bidi support, in `<>`, `[]()` by default.

**Comment by laukstein** [comment]
Custom HTTP headers compatibility for Github Pages (Webrick isn't planned to support https://github.com/github/pages-gem/issues/415)

**Comment by pathawks** [comment]
@laukstein Let’s keep this restricted to Jekyll features. We don’t have any control over GitHub Pages.

**Comment by letrastudio** [comment]
I'd love to see automatic generation of tag and category archive pages in core (plugins like [jekyll-archives](https://github.com/jekyll/jekyll-archives/) do this, but none are whitelisted in GitHub Pages). I wrote up #6952 with my detailed rationale and a proposal of how it could work.

Here's the TL;DR:

> Having vanilla Jekyll auto-generate archive pages for every tag and category is probably a good idea now, and I think I've come up with a good, clean way to do it while staying true to Jekyll's philosophy.

Code-wise, this snippet gets most of it across:

```yaml
collections:
  tags:
    output: true
    permalink: /tags/:name/
  categories:
    output: true
    permalink: /categories/:name/
```

**Comment by mbrav** [comment]
I know that this is markdown and not necessarily jekyll related, but my life would have been a thousand times easier if jekyll supported inline footnotes like so  ‘[^Footnote, p. 123]’ as apposed to the more tedious:

‘’’
This is first reference[^one], this is the second[^two]

[^one]: Footnote, p.1
[^two]: Footnote, p.2

‘’’

**Comment by DirtyF** [comment]
@mbrav This depedens on the Markdown engine. Please ask  on [Kramdown](https://github.com/gettalong/kramdown) or [CommonMark](http://commonmark.org/).

**Comment by asipple1** [comment]
First would like to say thank you to the Jekyll team for all the great work you all do. Jekyll is one my favorite frameworks to work with and I really do appreciate what you guys do! 
Ok on to the request it would be great if there was a built-in way to paginate a list Jekyll collection items on a page. I know there is [jekyll-paginate-v2 gem](https://github.com/sverrirs/jekyll-paginate-v2) but it would be nice if there was a built in pagniation that worked for both post and collection items.

**Comment by nativerez** [comment]
Hello all, my first post so go easy. I want to just say that I love Jekyll, I'm now using it on a lot of my web builds. One thing that has bothered me though, is that I've not been able to take advantage of the latest Jekyll features because a lot of the 3rd party CMS solutions I rely on for my clients use (cloudcannon, siteleaf and forestry) simply don't run the latest versions of Jekyll themselves. This forces me to have my folder structures in a really messy state and not take advantage of the latest features when developing locally. So if anyone at Jekyll could bend a few arms to get them to update that would be fab.

**Comment by DirtyF** [comment]
@nativerez It alwyas take some time to update to the latest version for services like GitHub Pages, Forestry, CloudCannon or Siteleaf because they need to run tests and adapt their tools. There's nothing Jekyll's core team can do about it.

**Comment by hosnas** [comment]
I would like to see [reading metadata in separate files](https://github.com/jekyll/jekyll/issues/1082).

**Comment by letrastudio** [comment]
@hosnas Great suggestion. I'll add a use case:
- Create a collection of static image files
- Add metadata like alt tags, captions, dimensions in sidecar files
- Loop through them to create a rich image gallery

Jekyll CMS services like [Siteleaf](https://siteleaf.com) could make great use of this feature. Siteleaf uploads static assets to an `_uploads` collection. cc @sskylar

**Comment by cmhelmr** [comment]
I would like to see the tags of collections included in site.tags. Thanks!

**Comment by BerkhanBerkdemir** [comment]
Please please support for i18n. At least 2 languages. Many plug-ins break or don't work with ghpages.

**Comment by shivajivarma** [comment]
build in support for content blocks and components

**Comment by kblife** [comment]
I'd very much love to have native support for Coffeescript in much the same manner as Sass. Thus, we could define in `_config.yml`

```
# Support Coffeescript
coffeescript:
  coffeescript_dir: _coffee
  output: compressed
```

This way one could create partials of `.coffee` in well-structured folders with all of them compiling minified to one `.js` file, even same way as you have `css/style.sass` with all the imports of individual Sass files.

**Comment by charlesrocket** [comment]
it would be great to have more interactions with github, like pulling profile information (contributions, avatar, ), releases data - and convert it for proper publication. this would be vital for software website to create team and download pages that would not require multiple code changes

**Comment by charlesrocket** [comment]
also background section could get some improvements - building website that will look good on regular displays and retina quiet challenging. creating multiple backgrounds for could be the solution to decrease loading time

**Comment by DirtyF** [comment]
@charlesrocket GitHub data should be part of GitHub-metadata plugin not Jekyll-core

**Comment by charlesrocket** [comment]
@DirtyF sorry, i just started with jekyll a week ago) this repo is on the list. at this time, i guess that this info could be pulled by `metadata`, but shouldn't jekyll convert it for better (more universal) usage?

**Comment by DirtyF** [comment]
@charlesrocket This is off-topic, but FYI GitHub-metadata already provides access to these data through `site.github` namespace.

**Comment by DirtyF** [comment]
@charlesrocket It looks like so, as Jekyll is fully agnostic when it comes what's get generated in the front-end, it just transforms files into HTML. You can use plugins like `jekyll-assets` and/or `jekyll-cloudinary` to help you deal with responsive images.

**Comment by charlesrocket** [comment]
@DirtyF thanks for the input! im looking for every jekyll plugin i can find, put majority is out of date and has no support, so it pushed me to propose these functions into core (not counting jekyll-originated plugins)

**Comment by ashmaroli** [comment]
(I) really love this ability to minimize comments on GitHub.. :sparkles:

**Comment by asipple1** [comment]
I would like to see an automated tool like Webpack integrated into Jekyll to handle Sass and Javascript. 
Some of the biggest reasons are auto-prefixing for Sass and the ability to use ES6 syntax that can automatically be polyfilled with babel. I think having an automated tool handle the sass and javascript bits may speed up the compile time of Jekyll.

**Comment by ashmaroli** [comment]
> I think having an automated tool handle the sass and javascript bits may speed up the compile time of Jekyll

How is that..??

**Comment by asipple1** [comment]
A big reason for me is when doing frontend work and the Sass folder has many nested folders and files Jekyll sass can be slow. I have found letting web pack handle the sass instead of Jekyll improves speed and developer happiness.

**Comment by nickjj** [comment]
I'd love to see Docker volume mounts work with Docker for Windows.

Rails, Flask, Phoenix and Node applications that I've worked with all work fine in Docker for Windows. You volume mount in the code, make a change, and in milliseconds your change is reflected and ready to be seen in a browser.

With Jekyll 3.8, this does not happen. The file change is reflected inside of the container, but the `jekyll serve` command doesn't regenerate the file. Using `--force_polling` also doesn't work.

Everything works fine under WSL but Jekyll is the only reason I have Ruby installed on my machine, mainly because every other app from every other language / framework works wonderfully with Docker.

**Comment by ashmaroli** [comment]
> I have found letting web pack handle the sass instead of Jekyll improves speed

Jekyll Core doesn't handle pre-processing css and javascript. We have two plugins (both are written in Ruby): `jekyll-sass-converter` (included in `jekyll` by default) and `jekyll-coffeescript`.

The speed improvements are probably due to the difference between `ruby-sass` and `node-sass`

**Comment by shivajivarma** [comment]
@oe : something like this 

```html
<html>
  <head>
    {% contentblock scripts %}
  </head>
  <body>
    <div class="main">
      {{ content }}
    </div>
    <div class="sidebar">
      {% contentblock sidebar %}
    </div>
  </body>
</html>
```

```html
---
layout: default
---

Here is my post content.

{% contentfor sidebar %}
* Some content
* in a markdown list
* with some {{ 'liquid' }} tags too!
{% endcontentfor %}
```

https://github.com/rustygeldmacher/jekyll-contentblocks

**Comment by BerkhanBerkdemir** [comment]
When I see the [repo](https://github.com/rustygeldmacher/jekyll-contentblocks), I though GitHub pages. It has similarity between. `_sidebar.md`, `_footer.md` etc..

**Comment by nkuehn** [comment]
I would love to see an offically supported page generator from data files. With the stellar rise of "headless" (= API only) SaaS CMS systems [static site generators ](url)have become the tool of choice for types of sites that they have not been considered for in the past. 

e.g. [contentful.com](https://contentful.com)  provides an [official jekyll plugin](https://www.contentful.com/ecosystem/jekyll/) that pulls all the CMS content into the Jekyll _data folder as yaml.  You can then use that content inside pages and include you've already created, but you can't actively create new pages from the SaaS CMS.  This is possible with other generators, but I'd like to stick with Jekyll due to your maturity and ecosystem. 

e.g. the site I'm maintaining is a mix of content maintained by techies directly in the jekyll git and content written and uploaded by less technical people in contentful. contentful triggers the redeploy on [netlify.com](https://www.netlify.com/)  via  webhook when a new article was published.    

Overall a great stack, just that we have to rely on a somewhat unmaintained hacky "page generator" jekyll plugin to stitch it together.

**Comment by geraldb** [comment]
One more suggestions (I'd be happy to help along) - The Jekyll Documention is fantastic and outstanding. But, what's wrong (or better how to make it more awesome)?

Currently the documentation is part of the jekyll repo <https://github.com/jekyll/jekyll/tree/master/docs> and the layouts, styling, build scripts etc. is part of the documentation too - it's all mixed together. What's wrong with that?

I'd say with its own dedicated markdown files in a repo it would be easier to contribute / edit / retarget etc. and with a "proper" docu (remote) theme  in its own repo it would be easy to reuse and change the theme too.   A while ago (dare I say years) I started to "clean up" as an example the documentation use "plain vanilla" markdown files in the manuscripts (book/documentation) format. See the Hyde Press Bookshelf live <http://hydepress.github.io> and all the source repos <https://github.com/hydepress>. For the suggestion for Jekyll 4.0 it would just be a  new theme repo and a new docs repo (with markdown only). Would be great to see some booklets / guides / tutorials that could be made easily from the new docs repos using some (selected) pages etc.   Let me (us) know what you think.

**Comment by iBug** [comment]
I'd like an additional YAML file that defines site-wide variables so other programs can generate the file safely (instead of appending to `_config.yml`).

It could be better if we can have something like `{% include vars.yml %}` in `_config.yml` but I don't think it's necessary. Just allow loading extra variables from another file, even if the file name is hard-coded, would be enough.

A typical use case would be when I write a custom script that pre-processes some file and show in the information generated site

**Comment by oracaoja** [comment]
Make internal linking easier. Like generating a JSON of internals links including anchors.

**Comment by letrastudio** [comment]
@iBug If I understand correctly, you could do this sort of thing using two (or more) config files. You can use this [build command flag](https://jekyllrb.com/docs/configuration/#build-command-options) to specify multiple config files:

```
--config FILE1[,FILE2,...]
```

> Settings in later files override settings in earlier files.

Granted, this option is only available as a command line flag, so it wouldn't work if you're using GitHub Pages to build your site.

**Comment by iBug** [comment]
@letrastudio Thanks for the response. Command line flags is surely not a problem for me because I use Travis CI to build my site.

**Comment by jdobres** [comment]
The ability to specify that collection items with the attribute `published: false` should still have their YAML data exposed in the backend `site.my_collection` object. Presently, setting `published: false` makes the data entirely inaccessible, even on the backend.

**Comment by tyler-insight** [comment]
Would love to see a standardized implementation of a service worker built into Jekyll utilizing the methods in [The offline cookbook](https://jakearchibald.com/2014/offline-cookbook/). A way of configuring it so that you could easily designate what the 'shell app' is so that those files load from offline cache first, and then designating dynamic content (such as blog posts) that load from network first and only cache so many entries (since caching 100 blog posts is ridiculous).

I realize these things can be done now, but finding a way to configure it to work perfectly with Jekyll would be amazing and save a lot of people the struggle of figuring out how to properly implement a service worker for it.

**Comment by y-sira** [comment]
I would be grateful if you could resolve #6410.

**Comment by alzeih** [comment]
I would like some nice to have plugin improvements:

- Include all enabled plugins in site.plugins, including those available from the Gemfile.
- Something like `{% if plugin "foo" %}...` for code is that is conditional on having certain plugins available. 
- Don't parse these conditional blocks unless the plugin is available, so code inside doesn't crash if the plugin isn't available.

**Comment by pathawks** [comment]
> Include all enabled plugins in site.plugins, including those available from the Gemfile.

Can you add a bit more detail here? How is this different from the current state?

**Comment by ashmaroli** [comment]
> Don't parse these conditional blocks unless the plugin is available, so code inside doesn't crash if the plugin isn't available.

@alzeih This cannot be implemented because Liquid is parsed and rendered by the `liquid` gem. Jekyll only provides the means to use and extend Liquid constructs for use in template files.

**Comment by alzeih** [comment]
> @alzeih This cannot be implemented because Liquid is parsed and rendered by the liquid gem. Jekyll only provides the means to use and extend Liquid constructs for use in template files.

Ah, I see.

This mentions using partials as a workaround: https://github.com/Shopify/liquid/issues/704#issuecomment-188321757 

Having all plugins available in `site.plugins` would allow for this.

**Comment by pathawks** [comment]
> > Don't parse these conditional blocks unless the plugin is available, so code inside doesn't crash if the plugin isn't available.
>
> @alzeih This cannot be implemented because Liquid is parsed and rendered by the liquid gem. Jekyll only provides the means to use and extend Liquid constructs for use in template files.

This would be such a useful feature, I wonder if we could find a workaround. Maybe instead of using the regular `{% if %}` tag, we create a new tag that acts like the `{% raw %}` tag (so Liquid does not parse the contents, and then _Jekyll_ will render the contents if the plugin is installed.

This is just an ideas list. Let’s not let implementation details get in the way at this stage. We are programmers; _anything_ is possible.

**Comment by BerkhanBerkdemir** [comment]
What do you think for [SanitizeHelper](https://stackoverflow.com/questions/50340509/escape-javascript-with-jekyll-redcarpet), @ashmaroli and @pathawks

**Comment by cetinajero** [comment]
Thus Jekyll is focused to create mostly frontend-only sites... it would be awesome to have a better way to organize our `Javascript` in partials and get an `assets/main.min.js` as a result.

This way we could place all our partials in a `js_dir` (defaults could be maybe `<source>/_js`) and specify the output style with a style option in the `_config.yml` file like:

```YAML
javascript:
    style: compressed
```

I know we have this with [CoffeScript](https://jekyllrb.com/docs/assets/#coffeescript), but [it is now fading away and `ES6` rising and raising](https://trends.google.com/trends/explore?date=today%205-y&q=coffescript,ES6) due to the fact that [`ES6` is an excellent modern CoffeeScript alternative](https://robots.thoughtbot.com/replace-coffeescript-with-es6).

# TL;DR

It'll be very cool to see **GitHub Pages sites with minified ES6 created from well structured javascript partials**.

**Comment by merlinpatt** [comment]
I would like to see a plugin that lets you loop through any directory's files. This could let you list, display, or include the files in the directory. This might look like

```
{% assign files = directory: 'some/path/to/stuff' %}
{% for file in files %}
  {{file.name}}
{% endfor %}

{% for file in files %}
  {% include_relative file.path %}
{% endfor %}
```

**Comment by merlinpatt** [comment]
@mmistakes The issue with that is that the files have to then be outputted into the main site.

My main use case here is that I would like to be able to loop over files in a directory that won't be outputted or accessible to the public.

For example, I have a slide deck split into various parts, which I then use `include_relative` to put back in the main file.

### Current method - repeated includes

Directory structure

```
react/
  react_includes/
    part1.html
    part2.html
    ...
  index.html
```

react/index.html

```
{% include_relative react_includes/part1.html %}
{% include_relative react_includes/part2.html %}
...
```

_config.yml

```
exclude: ['react/react_includes']
```

### Preferred method - loop over directory

But I have a lot of parts, and add to it / move it around a lot, so I'd rather do something like this:

```
{% assign files = directory: 'react_includes/' %}
{% for file in files %}
  {% include_relative file.path %}
{% endfor %}
```

By using this method, I could reorganize my files, add new ones, or delete them and I wouldn't have to change the slide deck itself.

**Comment by chrisfinazzo** [comment]
@tyler-insight, I have a PR [open][] for this in Minima, the default theme you get if you run `jekyll new`. In that case, I based my cache off of `lib/site_template`, although I probably should go through the Cookbook to make sure I've covered all bases.

[open]: https://github.com/jekyll/minima/pull/234

**Comment by cameronmcefee** [comment]
[I proposed this in the old Jekyll 4 wishlist thread](https://github.com/jekyll/jekyll/issues/5307#issuecomment-358385811), but it'd be really nice to be able to have schemas for data, collections, frontmatter, etc.

**Comment by tim-bradley** [comment]
I'd like to see Multiple Outputs baked in so for each .md file we can output the usual html as well as say amp and json files too. For example:  `layout: page, page-amp, page-json` referencing `page.html`, `page-amp.html`, and `page.json.json` layout files respectively.

It might also be nice to to be able to specify the same html page at different locations/paths.

**Comment by asierra** [comment]
User comments please, filterable and maybe with optional moderation.

**Comment by oe** [comment]
@asierra You can use Disqus or some similar software for that, it's way outside the scope of Jekyll as a _static_ site generator.

**Comment by PsiACE** [comment]
Could you consider supporting language switching by processing files in po format, or by some other means?

**Comment by CoolRunning** [comment]
Hello Jekyll team!

I love working with collections, but the current option `output: true|false` enforces generation of either all or none of the pages. Having more control about which item of the collection is output would be appreciated.

The current way of putting `published: false` suppresses output, but also makes the document inaccessible (it does not show up while looping through the collection).

Thank you for the good work so far!

**Comment by oncletom** [comment]
I'd say: **making it easier to filter content by date**.

At the moment, my understanding is posts/collection `date` field is coerced into a Ruby `Date` object. Comparison with `where_exp` is impossible without workaround (casting a variable in `_config.yml`). There is a bit of background about this in #6581.

```
{% assign max_date = '2018-06-24' | date %}
{% assign posts = site.posts | where_exp: 'p', 'p.date < max_date' %}
{% for post in posts %}
  ...
{% endfor %}
```

I see this from a Jekyll user standpoint and I don't know whether this sits better as a `where_exp` logic or as a `date` operator logic. 

Thanks for making Jekyll such a nice tool :-)

----

A +1 comment on @geraldb proposal: I find remote data to be quite useful to avoid monolithic Jekyll setup. My use case would be to reference sub-projects on a main Jekyll websites, from remote JSON file and Atom feed.

**Comment by Convincible** [comment]
Oh, and: the only thing that keeps me continually considering whether I should switch to Hugo is speed. Can we find ways to make Jekyll:
* Much faster at generating the site? (why is Hugo so many times faster? - IDK)
* Much better at doing it incrementally? (perhaps with manual dependency hints)

**Comment by oe** [comment]
@Convincible Hugo is much faster because it's written in natively compiled Go rather than interpreted Ruby. We're not planning to change that.

**Comment by Convincible** [comment]
@oe Ah, fair enough – but then, within the confines of interpreted Ruby, speed is on my wishlist! Mainly an issue when you get to 1,000+ page blogs/sites... I'd love to keep using Jekyll even here and beyond.

**Comment by Ryuno-Ki** [comment]
Speaking of Hugo … something like Shortcodes ([in Hugo](https://gohugo.io/content-management/shortcodes/#readout) or [in WordPress](https://codex.wordpress.org/Shortcode)) would writing Markdown easier to me, because I can defer HTML constructs to another file (for example, generating HTML from JSON via a given partial, but denote where it should show up in my Markdown).

Downside in my experience is, that the HTML looks broken once the plugin for the shortcode is removed … (because the shortcode is interpreted as text).

So Jekyll should offer _hooks_ for shortcodes. The implementation should happen in plugins.

**Comment by Ryuno-Ki** [comment]
Oh, and this wishlist reminded me on an „argument” I had with @pathawks over in https://github.com/jekyll/jekyll-sitemap/issues/88#issuecomment-361295866 a couple of months ago.

Would be handy to have a way to support internationalisation!

**Comment by Harrix** [comment]
Add the ability to place the post and its files in the same folder.
```
_posts
├─┬ 2018-07-08-post
│ ├── 2018-07-08-post.md
│ ├── img.png
│ ├── img2.jpg
│ └── file.pdf
└─┬ 2018-07-08-post2
  ├── index.md
  ├── img.png
  └── document.doc
```

**Comment by Convincible** [comment]
@Harrix Why can't you do this at present?

**Comment by ashmaroli** [comment]
@Convincible At present, static files within a `_posts` directory are ignored..

**Comment by mmistakes** [comment]
@Harrix FYI there's a plugin you can use that allows you to place static files in the same folder as the post.

https://nhoizey.github.io/jekyll-postfiles/

**Comment by ashmaroli** [comment]
> So leveraging the multi-thread capabilities of modern processors when generating pages is the suggestion.

@mbrav We can't leverage multiple cores because Ruby (the language Jekyll is written in), itself doesn't leverage them.

**Comment by pathawks** [comment]
@ashmaroli Ruby certainly does support [threads](https://ruby-doc.org/core-2.3.0/Thread.html). The trick is making sure that we only render things in parallel that **do not** depend on each other, and how can you know that two pages do not interfere with each other without rendering them?

**Comment by ashmaroli** [comment]
@pathawks Yes, Ruby does support threads. But because of the `Global Interpreter Lock (GIL)`, threads in Ruby are never actually run in parallel..

**Comment by pathawks** [comment]
> threads in Ruby are never actually run in parallel

TIL 😭:sob::sob:

**Comment by nickmccurdy** [comment]
What if we made templates pure, or at least atomic?

**Comment by merlinpatt** [comment]
@pathawks what's an example where two pages depend on each other? I can see this being an issue for includes or layouts but pages themselves should be independent of each other

**Comment by ashmaroli** [comment]
> templates pure, or at least atomic

Do elaborate further @nickmccurdy

**Comment by ashmaroli** [comment]
@merlinpatt For example, an `index.md` could have a `{% link random-page.md %}`

**Comment by mohsenkhanpour** [comment]
Proper i18n support. There are many plug-ins but none of them is supported by GitHub Pages.

**Comment by BerkhanBerkdemir** [comment]
@mohsenkhanpour https://github.com/jekyll/jekyll/issues/6948#issuecomment-386162200

**Comment by gsantner** [comment]
default shiped trim,ltrim,rtrim:  https://github.com/jekyll/jekyll/issues/153

**Comment by mohsenkhanpour** [comment]
PWA support out of the box. 
There are some Jekyll plug-ins that do this already and their workaround and implemention is dead simple.

**Comment by pathawks** [comment]
@mohsenkhanpour What would a Jekyll site look like as a Progressive Web App? Can you elaborate?

**Comment by mohsenkhanpour** [comment]
@pathawks I updated the initial comment/request with my thoughts on why this is a good thing to implement in the upcoming Jekyll 4.0.

**Comment by DirtyF** [comment]
To benefit from PWA features I'd recommend the use of [jekyll-pwa](https://github.com/lavas-project/jekyll-pwa )plugin  based on workbox 3 by Google.

**Comment by mohsenkhanpour** [comment]
@DirtyF [jekyll-pwa](https://github.com/lavas-project/jekyll-pwa) is more of a Service Worker thingy than `manifest.json` thingy. 😂 
I am talking about vanilla `manifest.josn` and not Service Workers.

**Comment by mbrav** [comment]
The ability to render includes only once. This [question on Stack](https://stackoverflow.com/questions/30523270/jekyll-generate-an-include-once-and-include-it-to-all-pages) demonstrates the idea. To demonstrate why this is a great idea, here is a scenario that I have: say you want to have a tag cloud in every one of your posts. You have `tag-cloud.html` in your `_includes`, which allows you to generate a tag cloud on every post so the user can have a better time navigating the site. Instead of generating a tag cloud for *every* post, it will be nice to have the ability to generate an include *only once* and then have it included in every post. In my case, this could decrease the generation of my blog from 1 minute to 4 seconds.

**Comment by ashmaroli** [comment]
@mbrav The plugin mentioned in the StackOverflow you linked to, `jekyll-include-cache` is the best option available out there.
However, there is a related proposal for getting similar support in Core: #7108 ..
..and another (very distantly) related proposal at #7136 

You may *subscribe* to the above PRs to stay notified about any developments on the proposals

**Comment by kenman345** [comment]
I was reading over the thoughts about Open Collective in the blog post [HERE](https://jekyllrb.com/news/2018/08/01/jekyll-sponsoring/)

Moving towards a means to maintain the list of plugins better might be to move those lists to yaml files and/or individual yaml files.

If each entry in a section of the list here: [Available Plugins](https://jekyllrb.com/docs/plugins/#available-plugins)
Then we can make it easier for maintenance. You would simply add a yaml file in the appropriate location that corresponds to your section. The file would have the fields needed to be listed. This would be `name`, `url`, `description`... 
Part of why that helps things is no merge conflicts even become possible. Refactoring that page becomes less impacted by people adding to the lists, and we can add in additional fields, like `repository_url` that can be used to track the repository of the project in question. When a new breaking change comes into a PR, we can potentially use a bot/script to pull in all those files and if the repository url exists, that it checks for a given pattern (and ambitiously, opens a PR for an automated fix). 

All this can potentially also be achieved with a yaml file per section, but we dont gain as much in terms of potential merge conflicts, but this is a fairly straightforward data set at that point, so this might not even be a concern. Opening hundreds of yaml files does have the potential of adding on time to generate the site. 

Side note: This also makes it extremely easy to keep the list alphabetically ordered.

I know this is not exactly a solution to the idea: "Create a comprehensive official plugin and theme directory site" but it does cause a shift in how we store the data we already have thats more inline with recommendations Jekyll users would be used to. It also makes it a lot easier to pull that data out into some other repository to make a dedicated official plugin directory whenever that task might be taken on.

**Comment by tordans** [comment]
I would love to see better environment support.

_For example site.url:_ Right now site.url has an implicit env pattern that is hard coded for development ([Source](https://github.com/jekyll/jekyll/pull/6270)). This is very implizit, hard to understand and remember and not extendable. #5142

I would love to see this extended. Example:

```yaml
url:
  production: 'https://www.example.com'
  development: 'http://www.example.test' # this could have the a fallback like today, see "source" above
  staging: 'https://staging.example.com' # this would be possible
```

The pattern could be:
When jekyll calls a variable, it checks for `<var-name>.<env-name>`. The fallback is `<var-name>`. So it does not matter if I configure just `url` order `url.<env-name>`.

(This is a copy of https://github.com/jekyll/jekyll/issues/5307#issuecomment-360737818)

**Comment by letrastudio** [comment]
@tordans You can set up all kinds of fallbacks by using multiple config files, as I mentioned in [this comment](https://github.com/jekyll/jekyll/issues/6948#issuecomment-388162247).

I've come across this use case myself — I use `_config.yml` for production values, and an additional `_config_dev.yml` for development overrides. When building locally I run:

```
jekyll serve --config _config.yml,_config_dev.yml
```

**Comment by letrastudio** [comment]
Is anyone working on i18n support? I noticed @oe's enthusiasm for it, and [a lot of interest](https://github.com/jekyll/jekyll/issues/6948#issuecomment-386162200). I think I can make a significant contribution at the proposal/spec stage.

I’ve been working on a multilingual Jekyll site, and I’ve come up with a pretty good plugin-free solution! It’s up at [openhousemacau.com](https://www.openhousemacau.com), hosted on GitHub Pages with Jekyll 3.7.3. It’s still in active development, but the i18n part is fully functional, and can support any number of locales (not just the two it has now).

It’s quite a complete solution, featuring sane (DRY) content management with fallbacks between locales, string translation, and date format localization. It even supports permalink localization, though we ended up not using it for this site. It works by using:

- Parallel collections for everything (`_posts` and `_posts_zh`, for example)
- A `locales` key in `_config.yml`, set up in a similar way to `collections`
- [Scope path glob patterns](https://jekyllrb.com/docs/configuration/#glob-patterns-in-front-matter-defaults) in front matter defaults
- Localized strings in `site.data`
- A liquid include for localizing dates
- Another liquid include for setting up a bunch of relationships and variables

The biggest issue is that the last liquid include gets called a lot (especially when looping through documents), so site builds get slow quickly. While I can’t open source the entire site, I could open source the i18n system by making a demo site, if there’s interest.

Throughout development I’ve kept Jekyll 4.0 in mind, trying to think about how a native solution could eliminate the problems I’ve come up against. I think I’m pretty qualified at this point to submit a complete top-to-bottom specification proposal for discussion. Should I do it?

**Comment by kenman345** [comment]
So This is just a thought, I can try and make an issue for it with more details and thought through functionality, but I have noticed a potential gap in functionality.

If you use the `_drafts` folder, or have a post that is ```unpublished` in your `_posts` folder, then you still need to move things around for your content, like images, to not get published to the generated site.

Since a lot of users use `/assets/img/` folders for the beginning position of images associated to a post, I was thinking that we could make a new special key to use in posts that you add to the frontmatter.

The key would be something like `asset_paths:` or `asset_path:` that we merge to an array, and it can define any of the assets that should not be published if they do not match up to any other posts that are being published. 

So if I have the following in my frontmatter:
```yaml
asset_path: /assets/imgs/vacation/July/4
```

Then the post when in the drafts folder would not copy over any of that folder located at ` /assets/imgs/vacation/July/4`.
However, if a post is in `_posts` that contains the same frontmatter key/value, then it would get published as it was included in a post that is getting published. This could be done for individual file levels too. 

The last part would be that unless included and exclusively used for an unpublished post, that the folder/files in question would perform /be acted on exactly as they already are now, where they will be included or excluded based on the `_config.yml` settings.

**Comment by tordans** [comment]
In Reply to https://github.com/jekyll/jekyll/issues/6948#issuecomment-414783157

@letrastudio I am aware of that solution, but it only causes new problems.

**The Szenario is:**
I have a config-files with 100 lines of config.
One, maybe two line of this config needs to be changed for different environments.

With your solution, I need to duplicate 99 lines in three files (dev, staging, production) and manually sync them every time I make a change. **This is bound to fail!** 

People will forget to sync the changes and the main idea of a staging system (to behave like the production system) will be lost.

**Other solutions would be**
a. The one I describe in https://github.com/jekyll/jekyll/issues/6948#issuecomment-414783157
b. Allowing more than one config, one general-config-file for all env, and one config-file for each env. This is similar to the way rails does it. The env-config will overwrite the general-config in case that is needed.
c. Your solution, but with the ability to "include" or "reference" other config files inside the one I call with the jekyll build command. For this I would say "build with staging-config" and inside staging config "use this 1 line and include/use all other lines from this other config-file".
d. …?

**Comment by mmistakes** [comment]
@tordans You don't have to duplicate 99 lines for each environment specific config file. When you specify multiple config files at build Jekyll daisy chains them from left to right.

Meaning you could have a production config with all settings, and then for your staging and dev configs just add the lines that are unique or change. Jekyll will use everything from the first config, and override whatever comes next. For example, say you have

**_config.yml** (production settings with all configs)

```
title: My Awesome Site
description: Powered by Jekyll
url: https://prod-domain.com
```

and then a staging specific config (e.g. `_config.staging.yml`

```
url: https://staging-domain.com
```

When you run `bundle exec jekyll build` you'd get the prod `url`, if you run:

```
bundle exec jekyll build --config _config.yml,_config.staging.yml
```

You'd get the staging `url` along with all the other variables set in `_config.yml`.

**Comment by kenman345** [comment]
Wanted to add a comment about what I noted in [Comment](https://github.com/jekyll/jekyll/issues/6948#issuecomment-415061483) that this is essentially a solution for what @Harrix requested, but a bit more segregated to the standards that we expect blogs to keep for folder/file structure. Instead of making a change to store the content with the post, this just allows you to specify it directly.

**Comment by oe** [comment]
@letrastudio So the more I've looked into implementing i18n in Core, the less feasible it became practically. There's huge performance drawbacks, even for sites that only use one language, the time spent rewriting a huge part of the Core internals would very likely be better spent working on other features, and we've garnered feedback from some big users of Jekyll that for them, performance improvements are more important than a huge new breaking feature. Obviously this shouldn't influence our decision as a project all that much, but it did provide us with insight into why a full i18n implementation in Core wouldn't be feasible. What I think we __could__ do is provide baseline APIs upon which a (maybe officially supported) plugin could operate. If you have any such ideas, feel free to shoot them my direction!

**Comment by letrastudio** [comment]
@oe I understand, and I feel the same way about performance. However, I am not completely discouraged! Wouldn't performance impact depend largely on how the feature is designed? I _think_ that my proposed solution wouldn't significantly impact performance for single-language sites (and it wouldn't break Jekyll 3 sites at all). Big disclaimer though: I am not a Ruby programmer so I could be completely wrong.

I think I'll write up my ideas anyway and post them as a new issue — even if they're not feasible for Core, I hope they'll contribute to the discussion, maybe even inspire some intrepid plugin developer. Or they might help spark some ideas for the baseline APIs you suggested (such low-level discussion is probably a bit out of my league).

And still, I've found that i18n is quite doable with vanilla Jekyll 3. IMHO content management can actually be better than with existing plugins, though it's a bit complex to work with on the Liquid side. Some small tweaks to make that easier might be worth implementing; I'd happily settle for making it feel like less of a hack.

**Comment by kevinSuttle** [comment]
What are the plans for (Dart) Sass support in Jekyll 4.0 now that the Ruby implementation is deprecated? I'm _very_ interested to hear the responses from the maintainers.

**Comment by DirtyF** [comment]
@kevinSuttle We'll rely on sassc implementation, that is currently [removing dependency to Ruby sass](https://github.com/sass/sassc-ruby/pull/85)

**Comment by kevinSuttle** [comment]
Awesome. Thank you @DirtyF!

**Comment by kvz** [comment]
It would be great if `include` would take preference over `exclude`. So that you could `exclude: *`, and include `include: [ homepage.html ]` and do fast iterations over a single page. This is also how e.g. `rsync` and other unix tools operate, and more useful than the (current) other way around, because by default. This was also reported in ticket https://github.com/jekyll/jekyll/issues/4791 but stalled as it would be a breaking change. Seems like 4.0 would be a perfect moment. What do you say @envygeeks?

**Comment by ashmaroli** [comment]
> It would be great if `include` would take preference over `exclude`

@kvz I've submitted a PR that should handle what you're looking for, as a side-effect.. Would you be able to give that branch a test-run..?
```ruby
# Gemfile
gem 'jekyll', git: 'https://github.com/jekyll/jekyll.git', ref: 'refs/pull/7188/head'
```
Feedback invited at the [PR's url](https://github.com/jekyll/jekyll/pull/7188)

**Comment by kvz** [comment]
@kvz:
> It would be great if `include` would take preference over `exclude`

@ashmaroli:
> I've submitted a PR that should handle what you're looking for

Just wanted to share that we confirmed [that that PR](https://github.com/jekyll/jekyll/pull/7188#issuecomment-420336100) solves the `include` vs `exclude` issue, so it would be very neat if that could land in 4.0 💟

**Comment by cameronmcefee** [comment]
(I suggest this without understanding what would actually be involved in doing it)

I'd love to see performance work done regarding [glob patterns in frontmatter defaults](https://jekyllrb.com/docs/configuration/front-matter-defaults/#glob-patterns-in-front-matter-defaults). I think it's a tremendously useful feature, but it's hampered by the fact that the more useful it is for a project, the more detrimental it is to build times.

I use this feature heavily in scenarios where I have a collection of documents which represents different variations of a type of content. For example, in the [Sentry marketing resources](https://sentry.io/_/resources/) "resources" is a collection, which is broken up into folders representing podcasts, videos, and pdfs. Each of those folders has it's own defaults that are appropriate for the given medium.

In smaller collections there isn't much of an issue, but on our docs site, where a collection is 250+ pages, using a wildcard added 2s to a .8s build.

**Comment by DirtyF** [comment]
Thanks to everyone for the feedback, that's already a lot of food on our plate, we won't be able to implement *all* your requests, but we'll definitively do our best to fulfill your expectations. We're currently focusing on performance. As we will need your help, we might pick some priority issues and offer rewards thanks to our Open Collective sponsors. Stay tuned.
